File: ts/src/session/runtime-session-timeline.ts
import type { RuntimeSessionEvent, RuntimeSessionEventLog } from "./runtime-events.js";
import { RuntimeSessionEventType } from "./runtime-events.js";
import { runtimeSessionIdForRun } from "./runtime-session-ids.js";
import {
  summarizeRuntimeSession,
  type RuntimeSessionReadStore,
  type RuntimeSessionSummary,
} from "./runtime-session-read-model.js";

export type RuntimeSessionTimelineItem =
  | RuntimeSessionPromptTimelineItem
  | RuntimeSessionChildTaskTimelineItem
  | RuntimeSessionGenericTimelineItem;

export interface RuntimeSessionTimeline {
  summary: RuntimeSessionSummary;
  items: RuntimeSessionTimelineItem[];
  item_count: number;
  in_flight_count: number;
  error_count: number;
}

export interface RuntimeSessionPromptTimelineItem {
  kind: "prompt";
  status: "in_flight" | "completed" | "failed";
  sequence_start: number;
  sequence_end: number | null;
  started_at: string;
  completed_at: string | null;
  role: string;
  cwd: string;
  prompt_preview: string;
  response_preview: string;
  error: string;
  request_id: string;
  prompt_event_id: string;
  response_event_id: string;
}

export interface RuntimeSessionChildTaskTimelineItem {
  kind: "child_task";
  status: "started" | "completed" | "failed";
  sequence_start: number;
  sequence_end: number | null;
  started_at: string;
  completed_at: string | null;
  task_id: string;
  child_session_id: string;
  worker_id: string;
  role: string;
  cwd: string;
  depth: number | null;
  max_depth: number | null;
  result_preview: string;
  error: string;
}

export interface RuntimeSessionGenericTimelineItem {
  kind: "event";
  sequence: number;
  event_id: string;
  event_type: string;
  timestamp: string;
  title: string;
  details: Record<string, string | number | boolean>;
}

export function buildRuntimeSessionTimeline(log: RuntimeSessionEventLog): RuntimeSessionTimeline {
  const items: RuntimeSessionTimelineItem[] = [];
  const openPrompts: RuntimeSessionPromptTimelineItem[] = [];
  const promptsByRequestId = new Map<string, RuntimeSessionPromptTimelineItem>();
  const promptsByEventId = new Map<string, RuntimeSessionPromptTimelineItem>();
  const childTasks = new Map<string, RuntimeSessionChildTaskTimelineItem>();

  for (const event of log.events) {
    switch (event.eventType) {
      case RuntimeSessionEventType.PROMPT_SUBMITTED: {
        const item = promptItemFromEvent(event);
        openPrompts.push(item);
        promptsByEventId.set(item.prompt_event_id, item);
        if (item.request_id) {
          promptsByRequestId.set(item.request_id, item);
        }
        items.push(item);
        break;
      }
      case RuntimeSessionEventType.ASSISTANT_MESSAGE: {
        const prompt = findPromptForResponse(event, {
          openPrompts,
          promptsByEventId,
          promptsByRequestId,
        });
        if (prompt) {
          completePromptItem(prompt, event);
        } else {
          items.push(genericItemFromEvent(event));
        }
        break;
      }
      case RuntimeSessionEventType.CHILD_TASK_STARTED: {
        const item = childTaskItemFromStartedEvent(event);
        for (const key of childTaskCorrelationKeysFromEvent(event)) {
          childTasks.set(key, item);
        }
        items.push(item);
        break;
      }
      case RuntimeSessionEventType.CHILD_TASK_COMPLETED: {
        const item = findChildTaskForCompletion(event, childTasks);
        if (item) {
          completeChildTaskItem(item, event);
        } else {
          items.push(genericItemFromEvent(event));
        }
        break;
      }
      default:
        items.push(genericItemFromEvent(event));
        break;
    }
  }

  return {
    summary: summarizeRuntimeSession(log),
    items,
    item_count: items.length,
    in_flight_count: items.filter(isInFlightItem).length,
    error_count: items.filter(isErrorItem).length,
  };
}

export function readRuntimeSessionTimelineById(
  store: RuntimeSessionReadStore,
  sessionId: string,
): RuntimeSessionTimeline | null {
  const log = store.load(sessionId);
  return log ? buildRuntimeSessionTimeline(log) : null;
}

export function readRuntimeSessionTimelineByRunId(
  store: RuntimeSessionReadStore,
  runId: string,
): RuntimeSessionTimeline | null {
  return readRuntimeSessionTimelineById(store, runtimeSessionIdForRun(runId));
}

function promptItemFromEvent(event: RuntimeSessionEvent): RuntimeSessionPromptTimelineItem {
  return {
    kind: "prompt",
    status: "in_flight",
    sequence_start: event.sequence,
    sequence_end: null,
    started_at: event.timestamp,
    completed_at: null,
    role: readString(event.payload.role),
    cwd: readString(event.payload.cwd),
    prompt_preview: preview(event.payload.prompt),
    response_preview: "",
    error: "",
    request_id: readString(event.payload.requestId),
    prompt_event_id: event.eventId,
    response_event_id: "",
  };
}

function findPromptForResponse(
  event: RuntimeSessionEvent,
  state: {
    openPrompts: RuntimeSessionPromptTimelineItem[];
    promptsByEventId: Map<string, RuntimeSessionPromptTimelineItem>;
    promptsByRequestId: Map<string, RuntimeSessionPromptTimelineItem>;
  },
): RuntimeSessionPromptTimelineItem | undefined {
  const requestId = readString(event.payload.requestId);
  const promptEventId = readString(event.payload.promptEventId);
  const prompt = (requestId ? state.promptsByRequestId.get(requestId) : undefined)
    ?? (promptEventId ? state.promptsByEventId.get(promptEventId) : undefined);
  if (!prompt && (requestId || promptEventId)) {
    return undefined;
  }
  const matchedPrompt = prompt ?? state.openPrompts[0];
  if (!matchedPrompt) {
    return undefined;
  }

  state.promptsByEventId.delete(matchedPrompt.prompt_event_id);
  if (matchedPrompt.request_id) {
    state.promptsByRequestId.delete(matchedPrompt.request_id);
  }
  const idx = state.openPrompts.indexOf(matchedPrompt);
  if (idx !== -1) {
    state.openPrompts.splice(idx, 1);
  }
  return matchedPrompt;
}

function completePromptItem(
  item: RuntimeSessionPromptTimelineItem,
  event: RuntimeSessionEvent,
): void {
  const error = readString(event.payload.error);
  const isError = readBoolean(event.payload.isError) || error !== "";
  item.status = isError ? "failed" : "completed";
  item.sequence_end = event.sequence;
  item.completed_at = event.timestamp;
  item.response_preview = preview(event.payload.text);
  item.error = error;
  item.response_event_id = event.eventId;
  item.role ||= readString(event.payload.role);
  item.cwd ||= readString(event.payload.cwd);
}

function childTaskItemFromStartedEvent(event: RuntimeSessionEvent): RuntimeSessionChildTaskTimelineItem {
  return {
    kind: "child_task",
    status: "started",
    sequence_start: event.sequence,
    sequence_end: null,
    started_at: event.timestamp,
    completed_at: null,
    task_id: readString(event.payload.taskId),
    child_session_id: readString(event.payload.childSessionId),
    worker_id: readString(event.payload.workerId),
    role: readString(event.payload.role),
    cwd: readString(event.payload.cwd),
    depth: readNullableNumber(event.payload.depth),
    max_depth: readNullableNumber(event.payload.maxDepth),
    result_preview: "",
    error: "",
  };
}

function findChildTaskForCompletion(
  event: RuntimeSessionEvent,
  childTasks: Map<string, RuntimeSessionChildTaskTimelineItem>,
): RuntimeSessionChildTaskTimelineItem | undefined {
  for (const key of childTaskCorrelationKeysFromEvent(event)) {
    const item = childTasks.get(key);
    if (item) return item;
  }
  return undefined;
}

function childTaskCorrelationKeysFromEvent(event: RuntimeSessionEvent): string[] {
  const keys: string[] = [];
  const childSessionId = readString(event.payload.childSessionId);
  if (childSessionId) keys.push(`child:${childSessionId}`);
  const taskId = readString(event.payload.taskId);
  const workerId = readString(event.payload.workerId);
  if (taskId && workerId) keys.push(`task:${taskId}:worker:${workerId}`);
  if (taskId) keys.push(`task:${taskId}`);
  if (keys.length === 0) keys.push(`event:${event.eventId}`);
  return keys;
}

function completeChildTaskItem(
  item: RuntimeSessionChildTaskTimelineItem,
  event: RuntimeSessionEvent,
): void {
  const error = readString(event.payload.error);
  const isError = readBoolean(event.payload.isError) || error !== "";
  item.status = isError ? "failed" : "completed";
  item.sequence_end = event.sequence;
  item.completed_at = event.timestamp;
  item.result_preview = preview(event.payload.result);
  item.error = error;
  item.child_session_id ||= readString(event.payload.childSessionId);
  item.worker_id ||= readString(event.payload.workerId);
  item.role ||= readString(event.payload.role);
  item.cwd ||= readString(event.payload.cwd);
}

function genericItemFromEvent(event: RuntimeSessionEvent): RuntimeSessionGenericTimelineItem {
  const details = eventDetails(event.payload);
  const titleDetails = eventTitleDetails(details);
  const detailText = Object.entries(titleDetails)
    .map(([key, value]) => `${key}=${String(value)}`)
    .join(" ");
  return {
    kind: "event",
    sequence: event.sequence,
    event_id: event.eventId,
    event_type: event.eventType,
    timestamp: event.timestamp,
    title: `${event.eventType}${detailText ? ` ${detailText}` : ""}`,
    details,
  };
}

function eventTitleDetails(
  details: Record<string, string | number | boolean>,
): Record<string, string | number | boolean> {
  const titleDetails: Record<string, string | number | boolean> = {};
  for (const key of [
    "command",
    "tool",
    "exitCode",
    "taskId",
    "childSessionId",
    "entryId",
    "entryCount",
    "components",
  ]) {
    const value = details[key];
    if (value !== undefined) titleDetails[key] = value;
  }
  return titleDetails;
}

function eventDetails(payload: Record<string, unknown>): Record<string, string | number | boolean> {
  const details: Record<string, string | number | boolean> = {};
  for (const key of [
    "role",
    "cwd",
    "command",
    "tool",
    "exitCode",
    "taskId",
    "childSessionId",
    "entryId",
    "entryCount",
    "components",
    "ledgerPath",
    "generation",
  ]) {
    const value = payload[key];
    if (typeof value === "string" && value !== "") details[key] = preview(value);
    if (typeof value === "number" || typeof value === "boolean") details[key] = value;
  }
  const commandName = aliasedDetail(payload, "commandName");
  if (details.command === undefined && commandName !== undefined) {
    details.command = commandName;
  }
  const toolName = aliasedDetail(payload, "toolName");
  if (details.tool === undefined && toolName !== undefined) {
    details.tool = toolName;
  }
  return details;
}

function aliasedDetail(
  payload: Record<string, unknown>,
  key: string,
): string | undefined {
  const value = payload[key];
  return typeof value === "string" && value !== "" ? preview(value) : undefined;
}

function isInFlightItem(item: RuntimeSessionTimelineItem): boolean {
  return (item.kind === "prompt" && item.status === "in_flight")
    || (item.kind === "child_task" && item.status === "started");
}

function isErrorItem(item: RuntimeSessionTimelineItem): boolean {
  return (item.kind === "prompt" && item.status === "failed")
    || (item.kind === "child_task" && item.status === "failed");
}

function preview(value: unknown, maxLength = 240): string {
  if (value === undefined || value === null) return "";
  const raw = typeof value === "string" ? value : JSON.stringify(value);
  const normalized = raw.replace(/\s+/g, " ").trim();
  return normalized.length > maxLength ? `${normalized.slice(0, maxLength - 3)}...` : normalized;
}

function readString(value: unknown): string {
  return typeof value === "string" ? value : "";
}

function readBoolean(value: unknown): boolean {
  return typeof value === "boolean" ? value : false;
}

function readNullableNumber(value: unknown): number | null {
  return typeof value === "number" ? value : null;
}

File: autocontext/src/autocontext/session/runtime_session_timeline.py
from __future__ import annotations

import json
from typing import Any

from autocontext.session.runtime_events import RuntimeSessionEvent, RuntimeSessionEventLog, RuntimeSessionEventType
from autocontext.session.runtime_session_ids import runtime_session_id_for_run
from autocontext.session.runtime_session_read_model import RuntimeSessionReadStore, summarize_runtime_session

RuntimeSessionTimelineItem = dict[str, Any]
RuntimeSessionTimeline = dict[str, Any]


def build_runtime_session_timeline(log: RuntimeSessionEventLog) -> RuntimeSessionTimeline:
    items: list[RuntimeSessionTimelineItem] = []
    open_prompts: list[RuntimeSessionTimelineItem] = []
    prompts_by_request_id: dict[str, RuntimeSessionTimelineItem] = {}
    prompts_by_event_id: dict[str, RuntimeSessionTimelineItem] = {}
    child_tasks_by_correlation_key: dict[str, RuntimeSessionTimelineItem] = {}

    for event in log.events:
        if event.event_type == RuntimeSessionEventType.PROMPT_SUBMITTED:
            item = _prompt_item_from_event(event)
            open_prompts.append(item)
            prompts_by_event_id[item["prompt_event_id"]] = item
            if item["request_id"]:
                prompts_by_request_id[item["request_id"]] = item
            items.append(item)
        elif event.event_type == RuntimeSessionEventType.ASSISTANT_MESSAGE:
            prompt = _find_prompt_for_response(
                event,
                open_prompts=open_prompts,
                prompts_by_request_id=prompts_by_request_id,
                prompts_by_event_id=prompts_by_event_id,
            )
            if prompt is None:
                items.append(_generic_item_from_event(event))
            else:
                _complete_prompt_item(prompt, event)
        elif event.event_type == RuntimeSessionEventType.CHILD_TASK_STARTED:
            item = _child_task_item_from_started_event(event)
            correlation_key = _child_task_correlation_key_from_item(item)
            if correlation_key:
                child_tasks_by_correlation_key[correlation_key] = item
            items.append(item)
        elif event.event_type == RuntimeSessionEventType.CHILD_TASK_COMPLETED:
            correlation_key = _child_task_correlation_key_from_event(event)
            child_item = child_tasks_by_correlation_key.get(correlation_key) if correlation_key else None
            if child_item is None:
                items.append(_generic_item_from_event(event))
            else:
                _complete_child_task_item(child_item, event)
        else:
            items.append(_generic_item_from_event(event))

    return {
        "summary": summarize_runtime_session(log),
        "items": items,
        "item_count": len(items),
        "in_flight_count": len([item for item in items if _is_in_flight_item(item)]),
        "error_count": len([item for item in items if _is_error_item(item)]),
    }


def read_runtime_session_timeline_by_id(
    store: RuntimeSessionReadStore,
    session_id: str,
) -> RuntimeSessionTimeline | None:
    log = store.load(session_id)
    return build_runtime_session_timeline(log) if log is not None else None


def read_runtime_session_timeline_by_run_id(
    store: RuntimeSessionReadStore,
    run_id: str,
) -> RuntimeSessionTimeline | None:
    return read_runtime_session_timeline_by_id(store, runtime_session_id_for_run(run_id))


def _prompt_item_from_event(event: RuntimeSessionEvent) -> RuntimeSessionTimelineItem:
    return {
        "kind": "prompt",
        "status": "in_flight",
        "sequence_start": event.sequence,
        "sequence_end": None,
        "started_at": event.timestamp,
        "completed_at": None,
        "role": _read_str(event.payload.get("role")),
        "cwd": _read_str(event.payload.get("cwd")),
        "prompt_preview": _preview(event.payload.get("prompt")),
        "response_preview": "",
        "error": "",
        "request_id": _read_str(event.payload.get("requestId")),
        "prompt_event_id": event.event_id,
        "response_event_id": "",
    }


def _find_prompt_for_response(
    event: RuntimeSessionEvent,
    *,
    open_prompts: list[RuntimeSessionTimelineItem],
    prompts_by_request_id: dict[str, RuntimeSessionTimelineItem],
    prompts_by_event_id: dict[str, RuntimeSessionTimelineItem],
) -> RuntimeSessionTimelineItem | None:
    request_id = _read_str(event.payload.get("requestId"))
    prompt_event_id = _read_str(event.payload.get("promptEventId"))
    prompt = (prompts_by_request_id.get(request_id) if request_id else None) or (
        prompts_by_event_id.get(prompt_event_id) if prompt_event_id else None
    )
    if prompt is None and (request_id or prompt_event_id):
        return None
    matched = prompt or (open_prompts[0] if open_prompts else None)
    if matched is None:
        return None

    prompts_by_event_id.pop(matched["prompt_event_id"], None)
    if matched["request_id"]:
        prompts_by_request_id.pop(matched["request_id"], None)
    if matched in open_prompts:
        open_prompts.remove(matched)
    return matched


def _complete_prompt_item(item: RuntimeSessionTimelineItem, event: RuntimeSessionEvent) -> None:
    error = _read_str(event.payload.get("error"))
    is_error = _read_bool(event.payload.get("isError")) or error != ""
    item["status"] = "failed" if is_error else "completed"
    item["sequence_end"] = event.sequence
    item["completed_at"] = event.timestamp
    item["response_preview"] = _preview(event.payload.get("text"))
    item["error"] = error
    item["response_event_id"] = event.event_id
    item["role"] = item["role"] or _read_str(event.payload.get("role"))
    item["cwd"] = item["cwd"] or _read_str(event.payload.get("cwd"))


def _child_task_item_from_started_event(event: RuntimeSessionEvent) -> RuntimeSessionTimelineItem:
    return {
        "kind": "child_task",
        "status": "started",
        "sequence_start": event.sequence,
        "sequence_end": None,
        "started_at": event.timestamp,
        "completed_at": None,
        "task_id": _read_str(event.payload.get("taskId")),
        "child_session_id": _read_str(event.payload.get("childSessionId")),
        "worker_id": _read_str(event.payload.get("workerId")),
        "role": _read_str(event.payload.get("role")),
        "cwd": _read_str(event.payload.get("cwd")),
        "depth": _read_nullable_number(event.payload.get("depth")),
        "max_depth": _read_nullable_number(event.payload.get("maxDepth")),
        "result_preview": "",
        "error": "",
    }


def _child_task_correlation_key_from_item(item: RuntimeSessionTimelineItem) -> str:
    return _child_task_correlation_key(item["task_id"], item["child_session_id"])


def _child_task_correlation_key_from_event(event: RuntimeSessionEvent) -> str:
    return _child_task_correlation_key(
        _read_str(event.payload.get("taskId")),
        _read_str(event.payload.get("childSessionId")),
    )


def _child_task_correlation_key(task_id: str, child_session_id: str) -> str:
    if child_session_id:
        return f"child_session_id:{child_session_id}"
    if task_id:
        return f"task_id:{task_id}"
    return ""


def _complete_child_task_item(item: RuntimeSessionTimelineItem, event: RuntimeSessionEvent) -> None:
    error = _read_str(event.payload.get("error"))
    is_error = _read_bool(event.payload.get("isError")) or error != ""
    item["status"] = "failed" if is_error else "completed"
    item["sequence_end"] = event.sequence
    item["completed_at"] = event.timestamp
    item["result_preview"] = _preview(event.payload.get("result"))
    item["error"] = error
    item["child_session_id"] = item["child_session_id"] or _read_str(event.payload.get("childSessionId"))
    item["worker_id"] = item["worker_id"] or _read_str(event.payload.get("workerId"))
    item["role"] = item["role"] or _read_str(event.payload.get("role"))
    item["cwd"] = item["cwd"] or _read_str(event.payload.get("cwd"))


def _generic_item_from_event(event: RuntimeSessionEvent) -> RuntimeSessionTimelineItem:
    details = _event_details(event.payload)
    detail_text = " ".join(f"{key}={value}" for key, value in _event_title_details(details).items())
    return {
        "kind": "event",
        "sequence": event.sequence,
        "event_id": event.event_id,
        "event_type": event.event_type.value,
        "timestamp": event.timestamp,
        "title": f"{event.event_type.value}{f' {detail_text}' if detail_text else ''}",
        "details": details,
    }


def _event_title_details(details: dict[str, str | int | float | bool]) -> dict[str, str | int | float | bool]:
    return {
        key: details[key]
        for key in ["command", "tool", "exitCode", "taskId", "childSessionId", "entryId", "entryCount", "components"]
        if key in details
    }


def _event_details(payload: dict[str, Any]) -> dict[str, str | int | float | bool]:
    details: dict[str, str | int | float | bool] = {}
    for key in [
        "role",
        "cwd",
        "command",
        "tool",
        "exitCode",
        "taskId",
        "childSessionId",
        "entryId",
        "entryCount",
        "components",
        "ledgerPath",
        "generation",
    ]:
        value = payload.get(key)
        if isinstance(value, str) and value:
            details[key] = _preview(value)
        elif isinstance(value, int | float) and not isinstance(value, bool):
            details[key] = value
        elif isinstance(value, bool):
            details[key] = value
    return details


def _is_in_flight_item(item: RuntimeSessionTimelineItem) -> bool:
    return bool((item["kind"] == "prompt" and item["status"] == "in_flight") or (
        item["kind"] == "child_task" and item["status"] == "started"
    ))


def _is_error_item(item: RuntimeSessionTimelineItem) -> bool:
    return bool((item["kind"] == "prompt" and item["status"] == "failed") or (
        item["kind"] == "child_task" and item["status"] == "failed"
    ))


def _preview(value: Any, max_length: int = 240) -> str:
    if value is None:
        return ""
    raw = value if isinstance(value, str) else json.dumps(value)
    normalized = " ".join(raw.split())
    return f"{normalized[: max_length - 3]}..." if len(normalized) > max_length else normalized


def _read_str(value: Any) -> str:
    return value if isinstance(value, str) else ""


def _read_bool(value: Any) -> bool:
    return value if isinstance(value, bool) else False


def _read_nullable_number(value: Any) -> int | float | None:
    return value if isinstance(value, int | float) and not isinstance(value, bool) else None
